text/0000-macro-metavar-expr.md

```diff
@@ -0,0 +1,332 @@
+single `$` token.
+
+This is also necessary for unambiguously defining repetitions in nested
+macros.  For example, this resolves [issue 35853], as the example in
+that issue can be expressed as:
+
+```
+macro_rules! foo {
+    () => {
+        macro_rules! bar {
+            ( $$( $$any:tt )* ) => { $$( $$any )* };
+        }
+    };
+}
+
+fn main() { foo!(); }
+```
+
+[issue 35853]: https://github.com/rust-lang/rust/issues/35853
+
+## Larger example
+
+For a larger example of these metavariable expressions in use, consider the
+following macro that operates over three nested repetitions:
+
+```
+macro_rules! example {
+    ( $( [ $( ( $( $x:ident )* ) )* ] )* ) => {
+        counts = (${count(x, 1)}, ${count(x, 2)}, ${count(x)})
+        nested:
+        $(
+            indexes = (${index()}/${length()})
+            counts = (${count(x, 1)}, ${count(x)})
+            nested:
+            $(
+                indexes = (${index(1)}/${length(1)}, ${index()}/${length()})
+                counts = (${count(x)})
+                nested:
+                $(
+                    indexes = (${index(2)}/${length(2)}, ${index(1)}/${length(1)}, ${index()}/${length()})
+                    ${ignore(x)}
+                )*
+            )*
+        )*
+    };
+}
+```
+
+Given this input:
+```
+    example! {
+        [ ( A B C D ) ( E F G H ) ( I J ) ]
+        [ ( K L M ) ]
+    }
+```
+
+The macro would expand to:
+```
+    counts = (2, 4, 13)
+    nested:
+        indexes = (0/2)
+        counts = (3, 10)
+        nested:
+            indexes = (0/2, 0/3)
+            counts = (4)
+            nested:
+                indexes = (0/2, 0/3, 0/4)
+                indexes = (0/2, 0/3, 1/4)
+                indexes = (0/2, 0/3, 2/4)
+                indexes = (0/2, 0/3, 3/4)
+            indexes = (0/2, 1/3)
+            counts = (4)
+            nested:
+                indexes = (0/2, 1/3, 0/4)
+                indexes = (0/2, 1/3, 1/4)
+                indexes = (0/2, 1/3, 2/4)
+                indexes = (0/2, 1/3, 3/4)
+            indexes = (0/2, 2/3)
+            counts = (2)
+            nested:
+                indexes = (0/2, 2/3, 0/2)
+                indexes = (0/2, 2/3, 1/2)
+        indexes = (1/2)
+        counts = (1, 3)
+        nested:
+            indexes = (1/2, 0/1)
+            counts = (3)
+            nested:
+                indexes = (1/2, 0/1, 0/3)
+                indexes = (1/2, 0/1, 1/3)
+                indexes = (1/2, 0/1, 2/3)
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds additional syntax to the language, that program authors must learn
+and understand.  We may not want to add more syntax.
+
+The author believes it is worth the overhead of new syntax, as even though
+there exist workarounds for obtaining the information if it's really needed,
+these workarounds are sometimes difficult to discover and naive
+implementations can significantly harm compiler performance.
```

Comment:
I agree that the workarounds for these are non-obvious, and that it's worth giving them well-known names.  But I think the jump to new syntax could be better-motivated in the RFC.

For example, why not `std::macro_utils::count!(...)` instead of `${count(...)}`?  If it can be written as a macro as @RustyYato showed, that would then leave it up to an implementation to choose whether to add special compiler code to optimize it or just decide that the binary matching trick is good enough.

(I suspect it won't be too hard to convince me that syntax is worth it, but I'd still like to see it addressed in the RFC text.)

Reply:
The workaround macros work by expanding to `count!($( $value )*)`: i.e. the compiler must generate a sequence by expanding the repetition, re-parsing it as part of the `count!` macro invocation, and then computing the length.  This is the redundant additional work that this RFC seeks to address. 

The reason for new syntax is that these expansions occur during macro transcription, rather than as their own macro expansions.  `${count(ident)}` would be transcribed directly to the literal count, whereas `count!(ident)` in a macro body would be transcribed to `count!(ident)` (there is no change as the transcriber has nothing to do - it doesn't peek inside macro invocations), at which point the information about what `ident` means is lost and the `count!` macro has no knowledge about what it is counting or what context it is counting it in.

Another way to think of metavariable expressions is as "macro transcriber directives".  You can then think of the macro transcriber as performing the following:
* `$var` => the value of `var`
* `$( ... )*` => a repetition
* `${ directive }` => a special transcriber directive
* `$$` => `$`

Perhaps describing it like this makes it a bit clearer that these are special things the transcriber to do (not necessarily limited to counts and indexes, but that is what this RFC focuses on).

We could special-case these macro invocations during transcription, but that feels like a worse solution.  It would make it harder to understand what the macro transcriber is going to do with arbitrary code if you don't remember all of the special macros that don't work like other macros.

(Conversely, I think there might be existing special macros that might have been better written as metavariable expressions if they had already existed.  While I haven't thought it through fully, `file!()`, `line!()` and `column!()` spring to mind as candidates).